pages/[chain]/collection/[contract].tsx

```diff
@@ -155,6 +163,94 @@ const CollectionPage: NextPage<Props> = ({ id, ssr }) => {
     fallbackData: initialTokenFallbackData ? [ssr.tokens] : undefined,
   })
 
+  useTokenUpdateStream(id as string, collectionChain.id, {
+    onClose: () => setSocketState(0),
+    onOpen: () => setSocketState(1),
+    onMessage: ({
+      data: reservoirEvent,
+    }: MessageEvent<ReservoirWebsocketIncomingEvent>) => {
+      let hasChange = false
+
+      const newTokens = [...tokens]
+      const price = NORMALIZE_ROYALTIES
+        ? reservoirEvent.data?.market?.floorAskNormalized?.price?.amount
+            ?.decimal
+        : reservoirEvent.data?.market?.floorAsk?.price?.amount?.decimal
+      const tokenIndex = tokens.findIndex(
+        (token) => token.token?.tokenId === reservoirEvent?.data.token.tokenId
+      )
+      const token = tokenIndex > -1 ? tokens[tokenIndex] : null
+      if (token) {
+        //if the token has dynamic pricing we need to abort, this isn't supported in the websocket
+        if (token?.market?.floorAsk?.dynamicPricing) {
+          return
+        }
+        newTokens.splice(tokenIndex, 1)
+      }
+
+      if (!price) {
+        if (token) {
+          const endOfListingsIndex = tokens.findIndex(
+            (token) => !token.market?.floorAsk?.price?.amount?.decimal
+          )
+          const newTokenIndex = endOfListingsIndex > -1 ? endOfListingsIndex : 0
+          newTokens.splice(newTokenIndex, 0, {
+            ...token,
+            market: {
+              floorAsk: {
+                id: undefined,
+                price: undefined,
+                maker: undefined,
+                validFrom: undefined,
+                validUntil: undefined,
+                source: {},
+              },
+            },
+          })
+          hasChange = true
+        }
+      } else {
+        let updatedToken = token ? token : reservoirEvent.data
+        updatedToken = {
+          ...updatedToken,
+          market: {
+            floorAsk: NORMALIZE_ROYALTIES
+              ? reservoirEvent.data.market.floorAskNormalized
+              : reservoirEvent.data.market.floorAsk,
+          },
+        }
+        if (tokens) {
+          //take into account sorting
+          let updatedTokenPosition = tokens.findIndex((token) => {
```

Comment:
if sorting by rarity, skip this part and set the updatedTokenPosition to be current position of the token (already presorted by rarity since that doesn't change when a listing is added)